lib/thrift/clients/binary_framed.ex

```diff
@@ -0,0 +1,274 @@
+  @type backoff_fn :: ((retry_count) -> backoff_ms)
+
+  @type data :: iolist | binary
+  @type tcp_opts :: [
+    timeout: integer,
+    send_timeout: integer,
+    backoff_calculator: backoff_fn
+  ]
+
+  @type genserver_call_options :: [
+    timeout: integer
+  ]
+
+  @type options :: [
+    tcp_opts: tcp_opts,
+    gen_server_opts: genserver_call_options
+  ]
+
+  defmodule State do
+    @type t :: %State{host: String.t,
+                      port: (1..65535),
+                      tcp_opts: BinaryFramed.tcp_opts,
+                      timeout: integer,
+                      sock: pid,
+                      retries: integer,
+                      backoff_calculator: BinaryFramed.backoff_fn}
+    defstruct host: nil,
+              port: nil,
+              tcp_opts: nil,
+              timeout: 5000,
+              sock: nil,
+              retries: 0,
+              backoff_calculator: nil
+  end
+
+  require Logger
+  use Connection
+
+  def init({host, port, opts}) do
+    tcp_opts = Keyword.get(opts, :tcp_opts, [])
+
+    backoff_calculator = Keyword.get(tcp_opts, :backoff_calculator, &calculate_backoff/1)
+    {timeout, tcp_opts} = Keyword.pop(tcp_opts, :timeout, 5000)
+
+    s = %State{host: to_host(host),
+               port: port,
+               tcp_opts: tcp_opts,
+               timeout: timeout,
+               backoff_calculator: backoff_calculator,
+               sock: nil}
+
+    {:connect, :init, s}
+  end
+
+  @spec start_link(String.t, (0..65535), options) :: GenServer.on_start
+  def start_link(host, port, opts) do
+    Connection.start_link(__MODULE__, {host, port, opts})
+  end
+
+  def close(conn), do: Connection.call(conn, :close)
+
+  def connect(_, %{sock: nil, host: host, port: port, tcp_opts: opts, timeout: timeout, retries: retries, backoff_calculator: backoff_calculator} = s) do
+    opts = opts
+    |> Keyword.merge(@immutable_tcp_opts)
+    |> Keyword.put_new(:send_timeout, 1000)
+
+    case :gen_tcp.connect(host, port, opts, timeout) do
+      {:ok, sock} ->
+        {:ok, %{s | sock: sock, retries: 0}}
+
+      {:error, _} ->
+        new_retries = retries + 1
+        backoff = backoff_calculator.(new_retries)
+
+        msg = "Failed to connect to #{host} (after #{new_retries + 1} attempts), retrying in #{backoff}ms."
+        if retries <= 5 do
+          Logger.info(msg)
+        else
+          Logger.warn(msg)
+        end
+
+        {:backoff, backoff, %{s | retries: new_retries}}
+    end
+  end
+
+  def disconnect(info, %{sock: sock} = s) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+      {:close, from} ->
+        Connection.reply(from, :ok)
+
+      {:error, :closed} ->
+        Logger.error("Connection closed")
+
+      {:error, reason} ->
+        reason = :inet.format_error(reason)
+        Logger.error("Connection error: #{reason}")
+    end
+    {:connect, :reconnect, %{s | sock: nil}}
+  end
+
+  @spec oneway(pid, data) :: :ok
+  def oneway(conn, data) do
```

Comment:
Could you add a `@doc` for this function? I know what a "oneway" service call is in terms of Thrift, but it's not obvious so me what this exported function does.